packages/block-editor/src/hooks/border-width.js

```diff
@@ -65,28 +52,41 @@ export const BorderWidthEdit = ( props ) => {
 		let borderPaletteColor = borderColor;
 
 		const hasZeroWidth = parseFloat( newWidth ) === 0;
+		const hadPreviousZeroWidth = parseFloat( width ) === 0;
 
 		// Setting the border width explicitly to zero will also set the
 		// border style to `none` and clear border color attributes.
-		if ( hasZeroWidth ) {
+		if ( hasZeroWidth && ! hadPreviousZeroWidth ) {
+			// Before clearing color and style selections, keep track of
+			// the current selections so they can be restored when the width
+			// changes to a non-zero value.
+			setColorSelection( borderColor );
+			setCustomColorSelection( customBorderColor );
+			if ( borderStyle !== 'none' ) {
+				setStyleSelection( borderStyle );
+			}
+
+			// Clear style and color attributes.
 			borderPaletteColor = undefined;
 			newStyle.border.color = undefined;
 			newStyle.border.style = 'none';
 		}
 
-		// Restore previous border style selection if width is now not zero and
-		// border style was 'none'. This is to support changes to the UI which
-		// change the border style UI to a segmented control without a "none"
-		// option.
-		if ( ! hasZeroWidth && borderStyle === 'none' ) {
-			newStyle.border.style = styleSelection;
-		}
-
-		// Restore previous border color selection if width is no longer zero
-		// and current border color is undefined.
-		if ( ! hasZeroWidth && borderColor === undefined ) {
-			borderPaletteColor = colorSelection?.name;
-			newStyle.border.color = colorSelection?.color;
+		if ( ! hasZeroWidth && hadPreviousZeroWidth ) {
```

Comment:
I don't think this wrapping conditional is required. The original checks each took into account whether:
- there is a non-zero width set
- the style/color had been cleared due to a previous zero width

Might be nicer to avoid the nested conditionals, what do you think?

Reply:
Personally I think keeping it makes it a little easier to read/understand when we expect this to run without having to reason through the logic.

If we get rid of this conditional, the `onChange` handler would work in theory but would rely on being run a single time. Strict mode [deliberately double-invokes](https://reactjs.org/docs/strict-mode.html#detecting-unexpected-side-effects) so this actually fails locally:

The first time `onChange` is called, the `borderColor` attribute is set to `undefined` and the temp `colorSelection` value is populated. The _second_ time it is called the first conditional on L59 is no longer true, but `borderColor` is already undefined, so L86 is hit and we immediately restore the color (and likewise for style). The result is that when you set width to `0`, it appears that color and style are not cleared.

Reply:
Fair point. Thanks for the explanation 👍